AwsClientTracing: Null pointer exception thrown for AwsAsyncClientBuilders

Stack trace includes:
```
Caused by: java.lang.NullPointerException: null
	at brave.instrumentation.aws.AwsClientTracing$TracingExecutorFactory.<init>(AwsClientTracing.java:69)
	at brave.instrumentation.aws.AwsClientTracing.build(AwsClientTracing.java:47)
```

This is because the `getClientConfiguration()` does not get initialized by default in the default builder (thus blows up when calling `clientConfiguration.getMaxConnections()` in `TracingExecutorFactory`), instead Amazon uses `AwsSyncClientParams` to provide defaults if any are missing right before they call the build method:

```
    @Override
    public final TypeToBuild build() {
        return configureMutableProperties(build(getSyncClientParams()));
    }
```

Temporary work-around: Have async builders include: `.withClientConfiguration(new ClientConfigurationFactory().getConfig());`

--- a/brave-instrumentation/aws-java-sdk-core/pom.xml
+++ b/brave-instrumentation/aws-java-sdk-core/pom.xml
@@ -64,5 +64,11 @@
       <artifactId>aws-java-sdk-dynamodb</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.github.stefanbirkner</groupId>
+      <artifactId>system-rules</artifactId>
+      <version>1.19.0</version>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 </project>

--- a/brave-instrumentation/aws-java-sdk-core/src/main/java/brave/instrumentation/aws/AwsClientTracing.java
+++ b/brave-instrumentation/aws-java-sdk-core/src/main/java/brave/instrumentation/aws/AwsClientTracing.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016-2018 The OpenZipkin Authors
+ * Copyright 2016-2019 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at

--- a/brave-instrumentation/aws-java-sdk-core/src/main/java/brave/instrumentation/aws/AwsClientTracing.java
+++ b/brave-instrumentation/aws-java-sdk-core/src/main/java/brave/instrumentation/aws/AwsClientTracing.java
@@ -16,6 +16,7 @@ package brave.instrumentation.aws;
 import brave.http.HttpTracing;
 import brave.propagation.CurrentTraceContext;
 import com.amazonaws.ClientConfiguration;
+import com.amazonaws.ClientConfigurationFactory;
 import com.amazonaws.client.builder.AwsAsyncClientBuilder;
 import com.amazonaws.client.builder.AwsClientBuilder;
 import com.amazonaws.client.builder.ExecutorFactory;
@@ -26,6 +27,9 @@ public final class AwsClientTracing {
   public static AwsClientTracing create(HttpTracing httpTracing) {
     return new AwsClientTracing(httpTracing); // no builder yet as we don't need it yet.
   }
+
+  static final ClientConfigurationFactory defaultClientConfigurationFactory =
+      new ClientConfigurationFactory();
 
   final HttpTracing httpTracing;
   final CurrentTraceContext currentTraceContext;
@@ -43,8 +47,12 @@ public final class AwsClientTracing {
     if (builder instanceof AwsAsyncClientBuilder) {
       ExecutorFactory executorFactory = ((AwsAsyncClientBuilder) builder).getExecutorFactory();
       if (executorFactory == null) {
+        ClientConfiguration clientConfiguration = builder.getClientConfiguration();
+        if (clientConfiguration == null) {
+          clientConfiguration = defaultClientConfigurationFactory.getConfig();
+        }
         ((AwsAsyncClientBuilder) builder).setExecutorFactory(
-            new TracingExecutorFactory(currentTraceContext, builder.getClientConfiguration())
+            new TracingExecutorFactory(currentTraceContext, clientConfiguration)
         );
       } else {
         ((AwsAsyncClientBuilder) builder).setExecutorFactory(

--- a/brave-instrumentation/aws-java-sdk-core/src/test/java/brave/instrumentation/aws/AwsClientTracingTest.java
+++ b/brave-instrumentation/aws-java-sdk-core/src/test/java/brave/instrumentation/aws/AwsClientTracingTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016-2018 The OpenZipkin Authors
+ * Copyright 2016-2019 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at

--- a/brave-instrumentation/aws-java-sdk-core/src/test/java/brave/instrumentation/aws/AwsClientTracingTest.java
+++ b/brave-instrumentation/aws-java-sdk-core/src/test/java/brave/instrumentation/aws/AwsClientTracingTest.java
@@ -23,6 +23,8 @@ import com.amazonaws.auth.AWSStaticCredentialsProvider;
 import com.amazonaws.auth.BasicAWSCredentials;
 import com.amazonaws.client.builder.AwsClientBuilder;
 import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDBAsync;
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDBAsyncClientBuilder;
 import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
 import com.amazonaws.services.dynamodbv2.model.AttributeValue;
 import java.util.Collections;
@@ -34,6 +36,7 @@ import org.junit.After;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
+import org.junit.contrib.java.lang.system.EnvironmentVariables;
 import org.junit.rules.TestRule;
 import org.junit.rules.TestWatcher;
 import org.junit.runner.Description;
@@ -46,6 +49,9 @@ public class AwsClientTracingTest {
   @Rule
   public MockDynamoDBServer dynamoDBServer = new MockDynamoDBServer();
 
+  @Rule
+  public EnvironmentVariables environmentVariables = new EnvironmentVariables();
+
   private BlockingQueue<Span> spans = new LinkedBlockingQueue<>();
   private AmazonDynamoDB client;
 
@@ -54,8 +60,10 @@ public class AwsClientTracingTest {
     Tracing tracing = tracingBuilder().build();
     HttpTracing httpTracing = HttpTracing.create(tracing);
     AmazonDynamoDBClientBuilder clientBuilder = AmazonDynamoDBClientBuilder.standard()
-        .withCredentials(new AWSStaticCredentialsProvider(new BasicAWSCredentials("access", "secret")))
-        .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(dynamoDBServer.url(), "us-east-1"));
+        .withCredentials(
+            new AWSStaticCredentialsProvider(new BasicAWSCredentials("access", "secret")))
+        .withEndpointConfiguration(
+            new AwsClientBuilder.EndpointConfiguration(dynamoDBServer.url(), "us-east-1"));
 
     client = AwsClientTracing.create(httpTracing).build(clientBuilder);
   }
@@ -92,6 +100,16 @@ public class AwsClientTracingTest {
 
     Span sdkSpan = spans.poll(100, TimeUnit.MILLISECONDS);
     assertThat(sdkSpan.name()).isEqualToIgnoringCase("aws-sdk");
+  }
+
+  @Test
+  public void buildingAsyncClientWithEmptyConfigDoesNotThrowExceptions() {
+    Tracing tracing = tracingBuilder().build();
+    HttpTracing httpTracing = HttpTracing.create(tracing);
+    environmentVariables.set("AWS_REGION", "us-east-1");
+
+    AmazonDynamoDBAsync asyncClient =
+        AwsClientTracing.create(httpTracing).build(AmazonDynamoDBAsyncClientBuilder.standard());
   }
 
   private MockResponse createDeleteItemResponse() {

